runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java

```diff
@@ -912,30 +922,81 @@ public void translateNode(
       TypeInformation<WindowedValue<KV<K, OutputT>>> outputTypeInfo =
           context.getTypeInfo(context.getOutput(transform));
 
-      TupleTag<KV<K, OutputT>> mainTag = new TupleTag<>("main output");
-      WindowDoFnOperator<K, InputT, OutputT> doFnOperator =
-          new WindowDoFnOperator<>(
-              reduceFn,
-              fullName,
-              (Coder) windowedWorkItemCoder,
-              mainTag,
-              Collections.emptyList(),
-              new DoFnOperator.MultiOutputOutputManagerFactory<>(mainTag, outputCoder),
-              windowingStrategy,
-              new HashMap<>(), /* side-input mapping */
-              Collections.emptyList(), /* side inputs */
-              context.getPipelineOptions(),
-              inputKvCoder.getKeyCoder(),
-              keySelector);
+      List<PCollectionView<?>> sideInputs = ((Combine.PerKey) transform).getSideInputs();
 
-      // our operator excepts WindowedValue<KeyedWorkItem> while our input stream
-      // is WindowedValue<SingletonKeyedWorkItem>, which is fine but Java doesn't like it ...
-      @SuppressWarnings("unchecked")
-      SingleOutputStreamOperator<WindowedValue<KV<K, OutputT>>> outDataStream =
-          keyedWorkItemStream
-              .transform(fullName, outputTypeInfo, (OneInputStreamOperator) doFnOperator)
-              .uid(fullName);
-      context.setOutputDataStream(context.getOutput(transform), outDataStream);
+      if (sideInputs.isEmpty()) {
+        TupleTag<KV<K, OutputT>> mainTag = new TupleTag<>("main output");
+        WindowDoFnOperator<K, InputT, OutputT> doFnOperator =
+            new WindowDoFnOperator<>(
+                reduceFn,
+                fullName,
+                (Coder) windowedWorkItemCoder,
+                mainTag,
+                Collections.emptyList(),
+                new DoFnOperator.MultiOutputOutputManagerFactory<>(mainTag, outputCoder),
+                windowingStrategy,
+                new HashMap<>(), /* side-input mapping */
+                Collections.emptyList(), /* side inputs */
+                context.getPipelineOptions(),
+                inputKvCoder.getKeyCoder(),
+                keySelector);
+
+        // our operator excepts WindowedValue<KeyedWorkItem> while our input stream
```

Comment:
"expects" (I realize you didn't add this comment here, but it shows up in the diff)

Reply:
fixed here and elsewhere.